upgrade.php

```diff
@@ -130,36 +130,90 @@ function getTableIndexes($table)
     */
 
     $server_utf8_variant = 'utf8';
-    if ($link->server_version >= 50503) {
-        $server_utf8_variant = 'utf8mb4';
-    }
 
-    echo pad(' - Convert database to UTF-8');
+    echo pad(' - Convert database to utf8');
     $sql = 'ALTER DATABASE `' . DB_NAME .
         '` CHARACTER SET = ' . $mysql_utf8_variant[$server_utf8_variant]['charset'] .
         ' COLLATE = ' . $mysql_utf8_variant[$server_utf8_variant]['collation'];
     executeQuery($sql);
 
+    // Truncate needed for VARCHAR field used as PRIMARY or FOREIGN KEY when using UTF-8mb4
+
+    // Table users
+    echo pad(' - Fix schema for username field in `users` table');
+    $sql = "ALTER TABLE `users` CHANGE `username` `username` VARCHAR( 191 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+
+    // Table spamscores
+    echo pad(' - Fix schema for user field in `spamscores` table');
+    $sql = "ALTER TABLE `spamscores` CHANGE `user` `user` VARCHAR( 191 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+
+    // Table user_filters
+    echo pad(' - Fix schema for username field in `user_filters` table');
+    $sql = "ALTER TABLE `user_filters` CHANGE `username` `username` VARCHAR( 191 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+    
+    // Revert back some tables to the right values due to previous errors in upgrade.php
+
+    // Table audit_log
+    echo pad(' - Fix schema for username field in `audit_log` table');
+    $sql = "ALTER TABLE `audit_log` CHANGE `user` `user` VARCHAR( 255 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+
+    // Table users
+    echo pad(' - Fix schema for password field in `users` table');
+    $sql = "ALTER TABLE `users` CHANGE `password` `password` VARCHAR( 255 ) DEFAULT NULL";
+    executeQuery($sql);
+
+    echo pad(' - Fix schema for fullname field in `users` table');
+    $sql = "ALTER TABLE `users` CHANGE `fullname` `fullname` VARCHAR( 255 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+
+    // Table mcp_rules
+    echo pad(' - Fix schema for rule_desc field in `mcp_rules` table');
+    $sql = "ALTER TABLE `mcp_rules` CHANGE `rule_desc` `rule_desc` VARCHAR( 100 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+
+    // Table autorelease
+    echo pad(' - Fix schema for msg_id field in `autorelease` table');
+    $sql = "ALTER TABLE `autorelease` CHANGE `msg_id` `msg_id` VARCHAR( 255 ) NOT NULL";
+    executeQuery($sql);
+
+    echo pad(' - Fix schema for uid field in `autorelease` table');
+    $sql = "ALTER TABLE `autorelease` CHANGE `uid` `uid` VARCHAR( 255 ) NOT NULL";
+    executeQuery($sql);
+
+    // Convert database to UTF-8mb4 if MySQL ≥ 5.5.3
+    if ($link->server_version >= 50503) {
+        $server_utf8_variant = 'utf8mb4';
+        echo pad(' - Convert database to ' . $server_utf8_variant . '');
```

Comment:
lines 190-194 should be outside `if ($link->server_version >= 50503)` (always convert to utf8)